pyfr/readers/cgns.py

```diff
@@ -0,0 +1,435 @@
+        )
+
+        if ptset_type.value not in [self.PointRange, self.ElementRange]:
+            raise RuntimeError('Only element range BC is supported')
+
+        self.lib.cg_boco_read(file, base, zone, idx + 1, bcrange, None)
+
+        return {'name': name.value.decode('utf-8'),
+                'range': list(i for i in bcrange)}
+
+    def nsections(self, zone):
+        file = zone['base']['file']
+        base = zone['base']['idx']
+        zone = zone['idx']
+
+        n = c_int()
+        self.lib.cg_nsections(file, base, zone, n)
+
+        return n.value
+
+    def section_read(self, zone, idx):
+        file = zone['base']['file']
+        base = zone['base']['idx']
+        zidx = zone['idx']
+
+        name = create_string_buffer(32)
+        etype, start, end, nbdry = c_int(), c_int(), c_int(), c_int()
+        pflag, cdim = c_int(), c_int()
+
+        self.lib.cg_section_read(
+            file, base, zidx, idx + 1, name, etype, start, end, nbdry, pflag
+        )
+
+        self.lib.cg_ElementDataSize(file, base, zidx, idx + 1, cdim)
+
+        return {'zone': zone, 'idx': idx + 1, 'dim': cdim.value,
+                'etype': etype.value, 'range': (start.value, end.value)}
+
+    def elements_read(self, sect, conn):
+        file = sect['zone']['base']['file']
+        base = sect['zone']['base']['idx']
+        zone = sect['zone']['idx']
+        idx = sect['idx']
+
+        self.lib.cg_elements_read(file, base, zone, idx,
+                                  conn.ctypes.data, None)
+
+
+class CGNSZoneReader(object):
+    # CGNS element types to PyFR type (petype) and node counts
+    _cgns_map = {
+        3: ('line', 2), 4: ('line', 3), 24: ('line', 4), 40: ('line', 5),
+        5: ('tri', 3), 6: ('tri', 6), 26: ('tri', 10), 42: ('tri', 15),
+        7: ('quad', 4), 9: ('quad', 9), 28: ('quad', 16), 44: ('quad', 25),
+        10: ('tet', 4), 11: ('tet', 10), 30: ('tet', 20), 47: ('tet', 35),
+        12: ('pyr', 5), 13: ('pyr', 14), 33: ('pyr', 30), 50: ('pyr', 55),
+        14: ('pri', 6), 16: ('pri', 18), 36: ('pri', 40), 53: ('pri', 75),
+        17: ('hex', 8), 19: ('hex', 27), 39: ('hex', 64), 56: ('hex', 125),
+        20: ('mixed',),
+    }
+
+    def __init__(self, cgns, base, idx):
+        self._cgns = cgns
+        zone = cgns.zone_read(base, idx)
+
+        # Read nodes
+        self.nodepts = self._read_nodepts(zone)
+
+        # Read bc
+        bc = self._read_bc(zone)
+
+        # Read elements
+        self.elenodes = elenodes = {}
+        self.pents = pents = {}
+
+        # Construct elenodes and physical entity
+        newp = 0
+        for idx in range(cgns.nsections(zone)):
+            elerng, elenode = self._read_element(zone, idx)
+
+            name = 'fluid'
+            for bcname, bcrng in bc.items():
+                if elerng[0] >= bcrng[0] and elerng[1] <= bcrng[1]:
+                    name = bcname
+
+            pent = pents.setdefault(name)
```

Comment:
pent = pents.setdefault(name, idx).  So long as the pents do not need to be consecutive this should clean up the code as it allows us to get rid of newp.